### PR TITLE
Faction/Role Hud Icons

### PR DESCRIPTION
**Pull request by DeityLink**
https://cdn.discordapp.com/attachments/343178619192868864/416403883498405889/hudiconsgood.webm

Alright, so it's a bit different from how it used to be. Previously, you'd automatically get a hud icon when you were a certain kind of antag, namely cultist, wizard, nukeop, etc. But now that antagonists have been split between their role and their faction, I judged that since the goal of those icons is to help team antags find each others, it should be controlled by the faction datum. They now have an hud_icons list() that can either remain empty (which means no icons, should you need a faction whose members don't automatically know each others) or have one or more strings matching the icon_states in role_HUD_icons.dmi
When drawing the icons, the faction checks the role's logo_state var. If said var is in hud_icons, it'll be drawn. If not, the first icon in the hud_icons list will be used instead. Useful for factions whose members can have subroles (such has nuke ops and their leaders, or revs and their heads, or later wizard and their apprentices, etc)
Lastly, previously if you were multiple sorts of antags at the same time, you'd see your icons stack, hiding each others. Now, if there are multiple factions using icons in-game, the icons will blink and cycle.

icons
*forgot to make the top dummy a nukeop but you get the idea*

### CODE REVIEW

File: code/datums/gamemode/factions/faction.dm [hunk 1]
```diff
@@ -116,6 +132,75 @@
...
+						if(istype(R_target.antag.current.loc,/obj/mecha))
+							imageloc = R_target.antag.current.loc
+						var/hud_icon = R_target.logo_state//the icon is based on the member's role
+						if (!(R_target.logo_state in hud_icons))
+							hud_icon = hud_icons[1]//if the faction doesn't recognize the role, it'll just give it a default one.
+						var/image/I = image('icons/role_HUD_icons.dmi', loc = imageloc, icon_state = hud_icon)
+						I.pixel_x = 20 * PIXEL_MULTIPLIER
+						I.pixel_y = 20 * PIXEL_MULTIPLIER
+						I.plane = ANTAG_HUD_PLANE
+						to_chat(world,"[hud_icon] pause_before=[offset * HUDICON_BLINKDURATION];pause_after=[HUDICON_BLINKDURATION*(factions_with_icons - 1 - offset)]")
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/faction.dm]: Debug message, please fix.

Comment (DeityLink) [reply code/datums/gamemode/factions/faction.dm]: AHHHHHHHHH

File: code/modules/mob/living/login.dm [hunk 2]
```diff
@@ -14,7 +14,11 @@
 	if(ticker && ticker.mode)
 		switch(ticker.mode.name)
 			if("sandbox")
-		/*		CanBuild()
+				CanBuild()
+
+		if (iscultist(src) || iswizard(src) || isnukeop(src))
```
Comment (MadmanMartian) [inline code/modules/mob/living/login.dm]: Have an IsInFaction check or something, so it's more modular than sticking a macro to see if the person is in 'one of these factions'

And don't have the IsInFaction check just look at every faction macro

File: code/datums/gamemode/factions/faction.dm [hunk 3]
```diff
@@ -116,6 +132,74 @@
 /datum/faction/proc/check_win()
 	return
 
+//updating every icons at the same time allows their animate() to be sync'd, so we can alternate the one on top without any additional proc calls.
+/proc/update_faction_icons()
+	if (!ticker || !ticker.mode)
+		return
+
+	var/offset = 0
+	for (var/datum/faction/F in ticker.mode.factions)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/faction.dm]: There has got to be a better way that
>Looping through every faction
>Having it call update_hud_icons
>>Within that, it loops through every faction again
>>Then loops through the members of that particular faction 3 times

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: code/datums/gamemode/factions/faction.dm [hunk 4]
```diff
@@ -35,11 +35,15 @@
 	var/datum/objective_holder/objective_holder
 	var/datum/role/roletype = /datum/role
 	var/logo_state = "synd-logo"
+	var/list/hud_icons = list()
 
 /datum/faction/New()
 	..()
 	objective_holder = new
 	objective_holder.faction = src
+	for (var/datum/faction/F in ticker.mode.factions)
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/faction.dm]: Have a list of factions from which you remove src and loop through that.
BYOND just does nothing when you ask to remove an item from a list and the list doesn't contain it.

File: code/datums/gamemode/factions/faction.dm [hunk 5]
```diff
@@ -116,6 +132,74 @@
...
+			if(R.antag && R.antag.current && R.antag.current.client)
+				for(var/image/I in R.antag.current.client.images)
+					if(I.icon_state in hud_icons)
+						R.antag.current.client.images -= I
+
+		//then re-add them
+		for(var/datum/role/R in members)
+			if(R.antag && R.antag.current && R.antag.current.client)
+				for(var/datum/role/R_target in members)
+					if(R_target.antag && R_target.antag.current)
```
Comment (ShiftyRail) [inline code/datums/gamemode/factions/faction.dm]: None of these procs sleep or spawn. What's stopping you from doing it all in one go ?

**Review by MadmanMartian — APPROVED** [review]

**Comment by FudgePucker** [comment]
This raises an interesting question, if you are a cult member and a rev member, do you kill a head of staff if they are a cult member?
Do you have more loyalty to the revolution or to narsie?

**Comment by DeityLink** [comment]
It's up to the player to decide. Actual guidelines will be given to players when the time comes.

**Pull request merged**